### PR TITLE
Enhance cache modes in resource loading

### PR DESCRIPTION
**Pull request by RandomShaper**
- Unify documentation, hoping to clear misconcepctions about about propagation of the cache mode across dependant loads.
- Clarify in docs that `CACHE_MODE_REPLACE` now also works on the main resource (from #87008).
- Add two recursive modes, counterparts of `CACHE_MODE_REPLACE` and `CACHE_MODE_IGNORE`, since it seems some need them (see #59669, #82830).
- Let resources, even loaded with one of the ignore-cache modes, get a path, which is useful for tools.

This would supersede #87759, addressing the concerns discussed there.

For context, quoting f8d03b98e7eeeb726af4f653f1fde8b063cbad14, where the original intent of `CACHE_MODE_REUSE` was stated (in essence, only affects subresources); bolds mine:
> Improve resource load cache
> -Added a new method in Resource: reset_state , used for reloading the same resource from disk
> -**Added a new cache mode "replace" in ResourceLoader, which reuses existing loaded sub-resources but resets their data from disk (or replaces them if they chaged type)**
> -Because the correct sub-resource paths are always loaded now, this fixes bugs with subresource folding or subresource ordering when saving.

Finally, it's a bit unfortunate that the cache modes are exposed in two places, which forces duplicate docs. The next compat breakage would be a nice opportunity for changing the exposed `ResourceFormatLoader.load()` to refer to the enum in `ResourceLoader`.

### CODE REVIEW

File: core/core_bind.h [hunk 1]
```diff
@@ -63,9 +63,11 @@ class ResourceLoader : public Object {
 	};
 
 	enum CacheMode {
-		CACHE_MODE_IGNORE, // Resource and subresources do not use path cache, no path is set into resource.
-		CACHE_MODE_REUSE, // Resource and subresources use patch cache, reuse existing loaded resources instead of loading from disk when available.
-		CACHE_MODE_REPLACE, // Resource and subresource use path cache, but replace existing loaded resources when available with information from disk.
+		CACHE_MODE_IGNORE,
+		CACHE_MODE_REUSE,
+		CACHE_MODE_REPLACE,
```
Comment (Mickeon) [inline core/core_bind.h:68]: Unless the prior comments were completely wrong I don't see why removing them outright.

File: core/io/resource_loader.h [hunk 2]
```diff
@@ -47,9 +47,11 @@ class ResourceFormatLoader : public RefCounted {
 
 public:
 	enum CacheMode {
-		CACHE_MODE_IGNORE, // Resource and subresources do not use path cache, no path is set into resource.
-		CACHE_MODE_REUSE, // Resource and subresources use patch cache, reuse existing loaded resources instead of loading from disk when available.
-		CACHE_MODE_REPLACE, // Resource and subresource use path cache, but replace existing loaded resources when available with information from disk.
+		CACHE_MODE_IGNORE,
```
Comment (Mickeon) [inline core/io/resource_loader.h:50]: I believe something like this is doable and it would ensure we at least know there's some duplication when reading code.
```suggestion
		CACHE_MODE_IGNORE = ResourceLoader.CACHE_MODE_IGNORE,
```
I may be **completely** mistaken.

File: doc/classes/ResourceLoader.xml [hunk 3]
```diff
@@ -138,13 +138,19 @@
 			The resource was loaded successfully and can be accessed via [method load_threaded_get].
 		</constant>
 		<constant name="CACHE_MODE_IGNORE" value="0" enum="CacheMode">
-			The resource is always loaded from disk, even if a cache entry exists for its path, and the newly loaded copy will not be cached. Instances loaded with this mode will exist independently.
+			Resource and subresources are neither retrieved from nor stored into the resource cache. External resources are loaded with [constant CACHE_MODE_REUSE].
```
Comment (Mickeon) [inline doc/classes/ResourceLoader.xml:141]: Mmmh, more to sink my teeth into. Yummy.
It's weird. The current descriptions in this PR are fairly hard to follow. Also worth noting that the class reference often mentions "from disk" when talking about non-volatile memory.

I do have a suggestion but as I am looking at both old descriptions and this PR's behavior I'm not sure how to word it better at the moment.

```suggestion
The resource and its subresources are [b]always[/b] loaded and stored to disk, and are not cached. A resource from a file is loaded from cache when possible. Otherwise, it is loaded from disk and cached. 
```

File: core/core_bind.h (hunk 1 above)
Comment (RandomShaper) [reply core/core_bind.h:68]: I removed them to keep the docs as the only source of truth. Well, actually two sources of truth, because due to how this enum is doubly-exposed, they have to be repeated. I had the impression that, when there's a natural place in the docs for some item, the unwritten rule is to document there instead of the source code. I can understand some will check the C++ enum, but if we keep explanations there, we'll already have three copies. Sounds like it's too easy for them to get out-of-sync over maintenance.

File: core/io/resource_loader.h (hunk 2 above)
Comment (RandomShaper) [reply core/io/resource_loader.h:50]: That's a good point. However, I think that ship already sailed. This enum is not the only one duplicated across the source code. If we want to embrace an idiom like that to send a signal that this is just a mirror image of another (which I believe is a good idea), I'd suggest making a codebase-wide pass editing all cases. Also, that would be a good thing to make it easier to search for them to hopefully remove all duplicates next time we can break compat.

File: doc/classes/ResourceLoader.xml (hunk 3 above)
Comment (RandomShaper) [reply doc/classes/ResourceLoader.xml:141]: I'm open-arms to better-worded descriptions. English is, as one can notice, not my mother tongue. Also, I was too focused on making them technically accurate that I may have not written the friendliest descriptions. That said, this suggestion of yours contains a mistake, when you say _stored to disk_. The resources are loaded _from disk_ and stored _(in)to cache_.

File: core/core_bind.h (hunk 1 above)
Comment (Mickeon) [reply core/core_bind.h:68]: That's true, but for very "deep" and somewhat important constants such as these, it's very unlikely their behavior "desyncs" with the documentation unnoticed.

But at the same time you _did have_ to make this PR in part because of that, so I can understand.

Comment (RandomShaper) [reply core/core_bind.h:68]: Let's see if we can get more opinions for the scales to tip one way or the other.

Comment (KoBeWi) [reply core/core_bind.h:68]: The current documentation was copied from these comments. I think it's fine to remove them now.

File: doc/classes/ResourceFormatLoader.xml [hunk 4]
```diff
@@ -99,10 +99,19 @@
...
+			Resource and subresources are neither retrieved from nor stored into the resource cache. External resources are loaded with [constant CACHE_MODE_REUSE].
 		</constant>
 		<constant name="CACHE_MODE_REUSE" value="1" enum="CacheMode">
+			Resource, subresources, and external resources are retrieved from cache if present, instead of loaded. If not present, they are loaded and stored into the cache. The same rules are propagated to external resources.
 		</constant>
 		<constant name="CACHE_MODE_REPLACE" value="2" enum="CacheMode">
+			Like [constant CACHE_MODE_REUSE], but both the main resource and its subresources, if already in the cache, have their data refreshed instead of being recreated as completely new instances.
+		</constant>
+		<constant name="CACHE_MODE_IGNORE_DEEP" value="3" enum="CacheMode">
+			Like [constant CACHE_MODE_IGNORE], but propagates recursively among loads of external resources.
```
Comment (KoBeWi) [inline doc/classes/ResourceFormatLoader.xml:111]: By external resources you mean dependencies? The description isn't very clear on that.

**Comment by KoBeWi** [comment]
I gave it a test and I don't get the difference between CACHE_MODE_REPLACE and CACHE_MODE_REPLACE_DEEP, and I don't see CACHE_MODE_REPLACE_DEEP actually used anywhere. I assumed that CACHE_MODE_REPLACE would reload a resource without reloading its external dependencies, but the dependencies get reloaded too. Same with CACHE_MODE_IGNORE.

File: core/io/resource_format_binary.cpp [hunk 5]
```diff
@@ -772,10 +772,12 @@ Error ResourceLoaderBinary::load() {
 			}
 
 			res = Ref<Resource>(r);
-			if (!path.is_empty() && cache_mode != ResourceFormatLoader::CACHE_MODE_IGNORE) {
-				r->set_path(path, cache_mode == ResourceFormatLoader::CACHE_MODE_REPLACE); //if got here because the resource with same path has different type, replace it
-			} else if (!path.is_resource_file()) {
-				r->set_path_cache(path);
+			if (!path.is_empty()) {
+				if (cache_mode != ResourceFormatLoader::CACHE_MODE_IGNORE) {
+					r->set_path(path, cache_mode == ResourceFormatLoader::CACHE_MODE_REPLACE); //if got here because the resource with same path has different type, replace it
```
Comment (KoBeWi) [inline core/io/resource_format_binary.cpp:777]:
```suggestion
					r->set_path(path, cache_mode == ResourceFormatLoader::CACHE_MODE_REPLACE); // If got here because the resource with same path has different type, replace it.
```

Comment (AThousandShips) [reply core/io/resource_format_binary.cpp:777]:
```suggestion
					r->set_path(path, cache_mode == ResourceFormatLoader::CACHE_MODE_REPLACE); // If we got here because the resource with same path has different type, replace it.
```

File: core/io/resource_format_binary.cpp [hunk 6]
```diff
@@ -1187,7 +1189,20 @@ Ref<Resource> ResourceFormatLoaderBinary::load(const String &p_path, const Strin
...
+			loader.cache_mode = p_cache_mode;
+			break;
+		case CACHE_MODE_IGNORE_DEEP:
+			loader.cache_mode = ResourceFormatLoader::CACHE_MODE_IGNORE;
+			break;
+		case CACHE_MODE_REPLACE_DEEP:
+			loader.cache_mode = ResourceFormatLoader::CACHE_MODE_REPLACE;
+			break;
+	}
+	loader.cache_mode_for_external = p_cache_mode;
```
Comment (KoBeWi) [inline core/io/resource_format_binary.cpp:1205]:
```suggestion
	switch (p_cache_mode) {
		case CACHE_MODE_IGNORE:
		case CACHE_MODE_REUSE:
		case CACHE_MODE_REPLACE:
			loader.cache_mode_for_external = CACHE_MODE_REUSE;
			loader.cache_mode = p_cache_mode;
			break;
		case CACHE_MODE_IGNORE_DEEP:
			loader.cache_mode_for_external = p_cache_mode;
			loader.cache_mode = ResourceFormatLoader::CACHE_MODE_IGNORE;
			break;
		case CACHE_MODE_REPLACE_DEEP:
			loader.cache_mode_for_external = p_cache_mode;
			loader.cache_mode = ResourceFormatLoader::CACHE_MODE_REPLACE;
			break;
	}
```
I assume it was supposed to be like this (https://github.com/godotengine/godot/pull/88664#issuecomment-1962430116).

File: core/io/resource_format_binary.cpp (hunk 5 above)
Comment (RandomShaper) [reply core/io/resource_format_binary.cpp:777]: Applied.

So, could we say that a PR is responsible for the code sytle of any line it touches, regardless how deep the change on it is? Would that be a good way of thinking about this?

File: doc/classes/ResourceFormatLoader.xml (hunk 4 above)
Comment (RandomShaper) [reply doc/classes/ResourceFormatLoader.xml:111]: Yes.

I inherited the jargon from the description in the comments and now I see it's not as clear as we may want. I'm amending the wording in the next push.

File: doc/classes/ResourceLoader.xml [hunk 7]
```diff
@@ -138,13 +138,19 @@
...
-			If a resource is cached, returns the cached reference. Otherwise it's loaded from disk.
+			Resource, subresources, and dependencies (external resources) are retrieved from cache if present, instead of loaded. Those not cached are loaded and then stored into the cache. The same rules are propagated recursively down the tree of dependencies (external resources).
 		</constant>
 		<constant name="CACHE_MODE_REPLACE" value="2" enum="CacheMode">
-			The resource is always loaded from disk, even if a cache entry exists for its path. The cached entry will be replaced by the newly loaded copy.
+			Like [constant CACHE_MODE_REUSE], but the cacke is checked for the main resource (the one requested to be loaded) as well as each of its subresources. Those already in the cache, as long as the loaded and cached types match, have their data refreshed from storage into the already existing instances. Othereise, they are recreated as completely new objects.
+		</constant>
+		<constant name="CACHE_MODE_IGNORE_DEEP" value="3" enum="CacheMode">
+			Like [constant CACHE_MODE_IGNORE], but propagated recursively down the tree of dependencies (external resources)
```
Comment (KoBeWi) [inline doc/classes/ResourceLoader.xml:150]:
```suggestion
			Like [constant CACHE_MODE_IGNORE], but propagated recursively down the tree of dependencies (external resources).
```
Same below.

**Review by KoBeWi — APPROVED** [review]

File: doc/classes/ResourceFormatLoader.xml [hunk 8]
```diff
@@ -99,10 +99,19 @@
 	</methods>
 	<constants>
 		<constant name="CACHE_MODE_IGNORE" value="0" enum="CacheMode">
+			Neither the main resource (the one requested to be loaded) nor any of its subresources are retrieved from cache nor stored into it. Dependencies (external resources) are loaded with [constant CACHE_MODE_REUSE].
 		</constant>
 		<constant name="CACHE_MODE_REUSE" value="1" enum="CacheMode">
+			Resource, subresources, and dependencies (external resources) are retrieved from cache if present, instead of loaded. Those not cached are loaded and then stored into the cache. The same rules are propagated recursively down the tree of dependencies (external resources).
 		</constant>
 		<constant name="CACHE_MODE_REPLACE" value="2" enum="CacheMode">
+			Like [constant CACHE_MODE_REUSE], but the cacke is checked for the main resource (the one requested to be loaded) as well as each of its subresources. Those already in the cache, as long as the loaded and cached types match, have their data refreshed from storage into the already existing instances. Othereise, they are recreated as completely new objects.
```
Comment (AThousandShips) [inline doc/classes/ResourceFormatLoader.xml:108]:
```suggestion
			Like [constant CACHE_MODE_REUSE], but the cache is checked for the main resource (the one requested to be loaded) as well as each of its subresources. Those already in the cache, as long as the loaded and cached types match, have their data refreshed from storage into the already existing instances. Otherwise, they are recreated as completely new objects.
```

File: doc/classes/ResourceFormatLoader.xml [hunk 9]
```diff
@@ -99,10 +99,19 @@
 	</methods>
 	<constants>
 		<constant name="CACHE_MODE_IGNORE" value="0" enum="CacheMode">
+			Neither the main resource (the one requested to be loaded) nor any of its subresources are retrieved from cache nor stored into it. Dependencies (external resources) are loaded with [constant CACHE_MODE_REUSE].
 		</constant>
 		<constant name="CACHE_MODE_REUSE" value="1" enum="CacheMode">
+			Resource, subresources, and dependencies (external resources) are retrieved from cache if present, instead of loaded. Those not cached are loaded and then stored into the cache. The same rules are propagated recursively down the tree of dependencies (external resources).
```
Comment (AThousandShips) [inline doc/classes/ResourceFormatLoader.xml:105]:
```suggestion
			Resources, subresources, and dependencies (external resources) are retrieved from cache if present, instead of loaded. Those not cached are loaded and then stored into the cache. The same rules are propagated recursively down the tree of dependencies (external resources).
```

File: doc/classes/ResourceLoader.xml [hunk 10]
```diff
@@ -138,13 +138,19 @@
 			The resource was loaded successfully and can be accessed via [method load_threaded_get].
 		</constant>
 		<constant name="CACHE_MODE_IGNORE" value="0" enum="CacheMode">
-			The resource is always loaded from disk, even if a cache entry exists for its path, and the newly loaded copy will not be cached. Instances loaded with this mode will exist independently.
+			Neither the main resource (the one requested to be loaded) nor any of its subresources are retrieved from cache nor stored into it. Dependencies (external resources) are loaded with [constant CACHE_MODE_REUSE].
 		</constant>
 		<constant name="CACHE_MODE_REUSE" value="1" enum="CacheMode">
-			If a resource is cached, returns the cached reference. Otherwise it's loaded from disk.
+			Resource, subresources, and dependencies (external resources) are retrieved from cache if present, instead of loaded. Those not cached are loaded and then stored into the cache. The same rules are propagated recursively down the tree of dependencies (external resources).
```
Comment (AThousandShips) [inline doc/classes/ResourceLoader.xml:144]:
```suggestion
			Resources, subresources, and dependencies (external resources) are retrieved from cache if present, instead of loaded. Those not cached are loaded and then stored into the cache. The same rules are propagated recursively down the tree of dependencies (external resources).
```

File: doc/classes/ResourceLoader.xml [hunk 11]
```diff
@@ -138,13 +138,19 @@
...
 		</constant>
 		<constant name="CACHE_MODE_REPLACE" value="2" enum="CacheMode">
-			The resource is always loaded from disk, even if a cache entry exists for its path. The cached entry will be replaced by the newly loaded copy.
+			Like [constant CACHE_MODE_REUSE], but the cacke is checked for the main resource (the one requested to be loaded) as well as each of its subresources. Those already in the cache, as long as the loaded and cached types match, have their data refreshed from storage into the already existing instances. Othereise, they are recreated as completely new objects.
```
Comment (AThousandShips) [inline doc/classes/ResourceLoader.xml:147]:
```suggestion
			Like [constant CACHE_MODE_REUSE], but the cache is checked for the main resource (the one requested to be loaded) as well as each of its subresources. Those already in the cache, as long as the loaded and cached types match, have their data refreshed from storage into the already existing instances. Otherwise, they are recreated as completely new objects.
```

**Review by AThousandShips — APPROVED** [review]

File: doc/classes/ResourceLoader.xml (hunk 10 above)
Comment (RandomShaper) [reply doc/classes/ResourceLoader.xml:144]: Changed the wording a bit.
